crates/typst/src/geom/gradient.rs

```diff
@@ -0,0 +1,1070 @@
+use std::f64::consts::{FRAC_PI_2, PI, TAU};
+use std::f64::{EPSILON, NEG_INFINITY};
+use std::fmt::{Debug, Write};
+use std::hash::Hash;
+
+use ecow::EcoVec;
+use typst_macros::{cast, func, scope, ty};
+use typst_syntax::{Span, Spanned};
+
+use super::color::Rgba;
+use super::*;
+use crate::diag::{bail, error, SourceResult};
+use crate::eval::{array, Array, Func, IntoValue};
+use crate::geom::{ColorSpace, Smart};
+
+/// A color gradient.
+///
+/// Typst supports:
+/// - Linear gradients through the [`gradient.linear` function]($gradient.linear)
+/// - 🚧 Radial gradient will soon™️ be available
+/// - 🚧 Conic gradient will soon™️ be available
+///
+/// ## Stops
+///
+/// A gradient is composed of a series of stops, each stop has a color and an offset.
+/// The offset is a [ratio]($ratio) between 0% and 100% that determines the position
+/// of the stop along the gradient. The color is the color of the gradient at that
+/// position. In Typst, you can chose to omit the offset, in which case it will be
+/// automatically computed for you, and all the stops will be evenly spaced.
+///
+/// ## Usage
+///
+/// Gradients can be used for the following purposes:
+/// - As fills to paint the interior of a shape: `rect(fill: gradient.linear(..))`
+/// - As strokes to paint the outline of a shape: `rect(stroke: 1pt + gradient.linear(..))`
+/// - As color maps you can [sample]($gradient.sample) from:
+///   `gradient.linear(..).sample(0.5)`
+///
+/// ## 🚧 Gradients on text
+///
+/// Currently gradients are not supported on text. However, in an upcoming release
+/// gradients will be supported on text.
+///
+/// ## Relative-ness
+///
+/// Gradients can be relative to either the shape they are painted on, or to the
+/// nearest parent containers. This is controlled by the `relative` argument of the
+/// constructors. By default, gradients are relative to the shape they are painted on,
+/// unless they are painted on text, in which case they are relative to the parent.
+///
+/// The way the parent is determined is as follows:
+/// - For shapes that are placed at the root/top level of the document, the parent
+///   is the page itself.
+/// - For other shapes, the parent is the first [`block`]($block) or [`box`]($box)
+///   that contains the shape. This includes the boxes and blocks that are implicitly
+///   created by show rules. For example, a [`rotate`]($rotate) will not affect the
+///   parent of a gradient, but a [`grid`]($grid) will.
```

Comment:
I think the idea of "first" here is not very clear when thinking of the document as a tree structure. Perhaps it would be better to clarify whether it is "outermost" or "innermost" (I think those terms are clearer regarding who is the parent of whom).